resources/testharness.js

```diff
@@ -1621,29 +1632,99 @@ policies and contribution forms [3].
-                        // within subsequent cleanup functions are not run.
-                        tests.phase = tests.phases.ABORTED;
-                        error_count += 1;
+                        on_error();
+                        return;
                     }
+
+                    results.push(result);
                 });
 
-        if (error_count > 0) {
-            total = this._user_defined_cleanup_count;
+        if (!this._is_promise_test) {
+            forEach(results,
+                    function(result) {
+                        if (result !== undefined) {
+                            bad_value_count += 1;
+                        }
+                    });
+            cleanup_done(this_obj, error_count, bad_value_count);
+        } else {
+            all_async(results,
+                      function(result, done) {
+                          if (result && typeof result.then === "function") {
+                              result
+                                  .then(null, on_error)
+                                  .then(done);
+                          } else {
+                              if (result !== undefined) {
+                                  bad_value_count += 1;
+                              }
+
+                              done();
+                          }
+                      },
+                      function() {
+                          cleanup_done(this_obj, error_count, bad_value_count);
+                      });
+        }
+    };
+
+    function cleanup_done(test, error_count, bad_value_count) {
+        if (error_count || bad_value_count) {
+            var total = test._user_defined_cleanup_count;
+
+            tests.phase = tests.phases.ABORTED;
+            forEach(tests.tests,
+                    function(test) {
+                        test.phase = test.phases.COMPLETE
+                    });
+
             tests.status.status = tests.status.ERROR;
-            tests.status.message = "Test named '" + this.name +
-                "' specified " + total + " 'cleanup' function" +
-                (total > 1 ? "s" : "") + ", and " + error_count + " failed.";
+            tests.status.message = "Test named '" + test.name +
+                "' specified " + total +
+                " 'cleanup' function" + (total > 1 ? "s" : "");
+
+            if (error_count) {
+                tests.status.message += ", and " + error_count + " failed";
+            }
+
+            if (bad_value_count) {
+                var type = test._is_promise_test ? "non-thenable " : "";
+                tests.status.message += ", and " + bad_value_count +
+                    " returned a " + type + "value";
+            }
+
+            tests.status.message += ".";
+
             tests.status.stack = null;
         }
-    };
+
+        test.phase = test.phases.COMPLETE;
+        tests.result(test);
+        forEach(test._done_callbacks,
+                function(cb) {
```

Comment:
Expand to `callback` as in most other places in this file.

Reply:
Definitely. I don't know what came over me!

Reply:
:)